### PR TITLE
fix windows service stop to be a clean exit

### PR DESCRIPTION
**Pull request by stiliev**
fix windows service stop to be a clean exit: not killing Agent.Listener, but passing CTRL-C for a graceful exit

### CODE REVIEW

File: src/Agent.Service/Windows/AgentService.cs [hunk 1]
```diff
@@ -145,11 +153,27 @@ protected override void OnStop()
...
+                            {
+                                //Generate console event for current console with GenerateConsoleCtrlEvent (processGroupId should be zero)
+                                GenerateConsoleCtrlEvent(CTRL_C_EVENT, 0);
+                                //Wait for the process to finish (give it up to 30 seconds)
+                                AgentListener.WaitForExit(30000);
+                            }
+                            finally
+                            {
+                                //Disconnect from console and restore Ctrl+C handling by main process
+                                FreeConsole();
```
Comment (ericsciple) [inline src/Agent.Service/Windows/AgentService.cs]: Is there any race condition when these functions are called? E.g. if child process has already exited vs not.

File: src/Agent.Service/Windows/AgentService.cs [hunk 2]
```diff
@@ -12,6 +13,7 @@ namespace AgentService
     public partial class AgentService : ServiceBase
     {
         public const string EventSourceName = "VstsAgentService";
+        private const int CTRL_C_EVENT = 0;
```
Comment (ericsciple) [inline src/Agent.Service/Windows/AgentService.cs]: Curious if you looked into sending sigterm instead?

Comment (stiliev) [inline src/Agent.Service/Windows/AgentService.cs]: Agent.Listener does not handle SIGTERM. I see that .Net Core is adding capability to handle SIGTERM in he last month, but have not tested it.

File: src/Agent.Service/Windows/AgentService.cs (hunk 1 above)
Comment (stiliev) [inline src/Agent.Service/Windows/AgentService.cs]: I think it is going to work fine if child process has exited. First we check "!AgentListener.HasExited" before running this code, then "AttachConsole" is going to fail if the child process has exited and nothing will be executed. If child process exits in the middle, then GenerateConsoleCtrlEvent will do nothing

Comment (ericsciple) [inline src/Agent.Service/Windows/AgentService.cs]: I forgot win32 apis don't throw and you need to check error explicitly.

looks good.

**Pull request merged**